libnvme-rs/src/nvme_uri.rs

```diff
@@ -0,0 +1,277 @@
+                source: ParseError::RelativeUrlWithCannotBeABaseBase,
+            }),
+            Some(s) => {
+                let segments = s.collect::<Vec<&str>>();
+                if segments[0].is_empty() {
+                    Err(NvmeError::UrlError {
+                        source: ParseError::RelativeUrlWithCannotBeABaseBase,
+                    })
+                } else {
+                    Ok(segments[0].to_string())
+                }
+            }
+        }?;
+
+        Ok(Self {
+            trtype,
+            traddr,
+            trsvcid: url.port().unwrap_or(4420),
+            subsysnqn: subnqn,
+        })
+    }
+}
+
+impl NvmeTarget {
+    /// Connect to NVMe target
+    /// Returns block device paths of NVMe namespaces that were connected
+    pub fn connect(&self) -> Result<Vec<String>, NvmeError> {
+        let r = NvmeRoot::new(unsafe { crate::nvme_scan(std::ptr::null()) });
+        let hostnqn =
+            NvmeStringWrapper::new(unsafe { crate::nvmf_hostnqn_from_file() });
+        let hostid =
+            NvmeStringWrapper::new(unsafe { crate::nvmf_hostid_from_file() });
+
+        let h = unsafe {
+            crate::nvme_lookup_host(
+                r.as_mut_ptr(),
+                hostnqn.as_ptr(),
+                hostid.as_ptr(),
+            )
+        };
+        if h.is_null() {
+            return Err(NvmeError::LookupHostError {
+                rc: -libc::ENOMEM,
+            });
+        }
+
+        let subsysnqn = std::ffi::CString::new(self.subsysnqn.clone()).unwrap();
+        let transport = std::ffi::CString::new(self.trtype.to_str()).unwrap();
+        let traddr = std::ffi::CString::new(self.traddr.clone()).unwrap();
+        let host_traddr = std::ptr::null();
+        let host_iface = std::ptr::null();
+        let trsvcid = std::ffi::CString::new(self.trsvcid.to_string()).unwrap();
+        let nvme_ctrl = unsafe {
+            crate::nvme_create_ctrl(
+                subsysnqn.as_ptr(),
+                transport.as_ptr(),
+                traddr.as_ptr(),
+                host_traddr,
+                host_iface,
+                trsvcid.as_ptr(),
+            )
+        };
+        if nvme_ctrl.is_null() {
+            return Err(NvmeError::CreateCtrlrError {
+                rc: -libc::ENOMEM,
+            });
+        }
+        let cfg = crate::nvme_fabrics_config {
+            queue_size: 0,
+            nr_io_queues: 0,
+            reconnect_delay: 0,
+            ctrl_loss_tmo: crate::NVMF_DEF_CTRL_LOSS_TMO as i32,
+            fast_io_fail_tmo: 0,
+            keep_alive_tmo: 0,
+            nr_write_queues: 0,
+            nr_poll_queues: 0,
+            tos: -1,
+            duplicate_connect: false,
+            disable_sqflow: false,
+            hdr_digest: false,
+            data_digest: false,
+        };
+        let ret = unsafe {
+            crate::nvmf_add_ctrl(h, nvme_ctrl, &cfg, cfg.disable_sqflow)
+        };
```

Comment:
What is the contract here? ret == 0 means we have, connected to the controller -- and its namespaces? or just the controller?

Reply:
This means it has successfully written to `/dev/nvme-fabrics` and read a response back, which AIUI just means it has connected to the controller.

Reply:
I just looked into the function and it does a whole lot more than that. Perhaps we are looking at different versions but master seems to add the controller (or returns <0) and then goes on and configure it and adds the controller to the subsystem. From there,  it seems to me we can use nvme_ctrL-first_ns() would be a lot better than this manual looping.

Reply:
`NvmeNamespaceIterator::next()` essentially calls `nvme_subsystem_first_ns()` or `nvme_subsystem_next_ns()`.

Reply:
Then I dont understand why the loop within intervals of 1 second, creates a new root and iterates over it. Once the controller is connected it is part of the root tree, and so we can just list the namespaces once, right?

Reply:
AIUI the kernel connects to the namespace asynchronously after connecting to the controller. We have a similar race condition in one of the unit tests which uses nvme-cli where a `nvme list` immediately after `nvme connect` returns no NVMe devices.